### ISSUE TITLE
Why do I get a QH when I optimize for a QA?

### ISSUE THREAD

**Issue by rahulgaur104** [PlasmaControl/DESC#1953]
Here's a minimally working script with the latest version of DESC, using which I am trying to get a simple QA equilibrium.

I start with @f0uriest 's generalized parametrization and optimize for QA and some shape-related quantities. It takes about 10 min on a 40G A100 GPU.
 
Even though the maximum Quasisymmetry error goes from  1e-1 -> 1e-14 the final quasisymmetry clearly close to a helical symmetry even though I specify helicity = (1, 0). I have tried both QuasisymmetryTwoTerm and QuasisymmetryBoozer.

Is there a bug in the Quasisymmetry classes? Am I missing something?

```python3

#!/usr/bin/env python3

from desc import set_device

set_device("gpu")

import jax

import numpy as np
from desc.backend import jnp, jax
from desc.geometry import FourierRZToroidalSurface
from desc.equilibrium import EquilibriaFamily, Equilibrium
from desc.grid import LinearGrid, Grid
from desc.objectives import (
    ForceBalance,
    ObjectiveFunction,
    MagneticWell,
    FixBoundaryR,
    FixBoundaryZ,
    FixCurrent,
    FixPressure,
    FixPsi,
    AspectRatio,
    FixPressure,
    Elongation,
    RotationalTransform,
    BScaleLength,
    QuasisymmetryTwoTerm,
    QuasisymmetryBoozer,
)
from desc.optimize import Optimizer
from desc.profiles import PowerSeriesProfile, SplineProfile
from desc.magnetic_fields import OmnigenousField
from desc.vmec import VMECIO
from desc.plotting import *

from desc.integrals import Bounce2D
from desc.compat import rescale, flip_theta

from matplotlib import pyplot as plt
import os
import sys


NFP = 1
aspect_ratio = 7.0
twist = 1.0
torsion = 0.0
elongation = 3.0
triangularity = -0.5
squareness = 0.1
eccentricity = 0.0


mirror_ratio = 0.2
well_weight = 2.0

# setting these values manually
helicity = (1, 0)
L = int(10)
M = int(10)
N = int(10)


# dependent variables
M_grid = int(np.ceil(2.5 * M))
N_grid = int(np.ceil(2.5 * N))

R0 = 1.0
minor_radius = R0 / aspect_ratio
flux = np.pi * minor_radius**2

init_fname = f'NFP{NFP}_A{aspect_ratio:+.1f}_twist{twist:+.1f}_tor{torsion:+.1f}_elong{elongation:+.1f}_tri{triangularity:+.1f}_square{squareness:+.1f}_ecc{eccentricity:+.1f}'
fname = init_fname

p_l = [0, 0., 0., 0., -0., 0.]

Ntheta = 100
Nphi = 200

theta = np.linspace(0, 2*np.pi, Ntheta)
phi = np.linspace(0, 2*np.pi, Nphi)

theta_2D, phi_2D = np.meshgrid(theta, phi, indexing="ij")

surf = FourierRZToroidalSurface.from_shape_parameters(major_radius=R0, aspect_ratio=aspect_ratio, elongation=elongation, triangularity=triangularity, squareness=squareness, eccentricity=eccentricity, torsion=torsion,  twist=twist, NFP=NFP)
flux = np.pi * minor_radius ** 2

eq = Equilibrium(Psi=flux, NFP=NFP, L=L, M=M, N=N, sym=True, surface=surf)
eq, _ = eq.solve(objective="force", ftol=1e-4, xtol=1e-6, gtol=1e-6, maxiter=120)

eq.save("eq_" + init_fname + "_vacuum_init_QS.h5")

print(eq.compute("<beta>_vol")["<beta>_vol"])  

rho = np.linspace(0, 1, L+1)
grid = LinearGrid(L=L, M = M_grid, N=N_grid, NFP=eq.NFP, sym=True)
data_keys = ["iota", "D_Mercier", "magnetic well"]
data = eq.compute(data_keys, grid=grid)

print("D_Mercier=", grid.compress(data["D_Mercier"]));
print("magnetic well=", grid.compress(data["magnetic well"]));
print("iota", grid.compress(data["iota"]))

fname = init_fname

surfaces_omni = [0.25, 0.4, 0.6, 0.85] 
QS_weights = [20, 40.0, 60, 80, 100, 150]
mirror_weights = [4e2, 7e2, 1e3, 2e3]
scale_weights = [4e2, 4e2, 8e2, 1e3]


start_idx = 1
gap = 1

staging_array = [1, 2, 3, 4]

for k in staging_array:
   
    print("\n---------------------------------------")
    print(f"Optimizing boundary modes M, N <= {k}")
    print("---------------------------------------")
    

    try:
        QS_weight = QS_weights[int(k-1)]
    except:
        QS_weight = QS_weights[-1]

    try:
        mirror_weight = mirror_weights[int(k-1)]
    except:
        mirror_weight = mirror_weights[-1]
   
    try:
        scale_weight = scale_weights[int(k-1)]
    except:
        scale_weight = scale_weights[-1]


    Elongation_grid = LinearGrid(M = 2*int(M), N = 2*int(N), rho=np.array([1.0]), NFP=eq.NFP, sym=False, axis=False)
    iota_grid = LinearGrid(M = 2*int(M), N = 2*int(N), rho=np.array([0.05]), NFP=eq.NFP, sym=True, axis=False)
    QS_grid = LinearGrid(
        M=M_grid, N=eq.N_grid, NFP=eq.NFP, rho=surfaces_omni, sym=False
    )

    modes_R = np.vstack(
        (
            [0, 0, 0],
            eq.surface.R_basis.modes[
                np.max(np.abs(eq.surface.R_basis.modes), 1) > k, :
            ],
        )
    )
    modes_Z = eq.surface.Z_basis.modes[
        np.max(np.abs(eq.surface.Z_basis.modes), 1) > k, :
    ]
    
    objective = ObjectiveFunction((ForceBalance(eq=eq, weight = 1e4),
                                   AspectRatio(eq=eq, bounds=(aspect_ratio-2.5, aspect_ratio+1.0), weight=1e4),
                                   Elongation(eq=eq, bounds=(3.0, 8.), grid=Elongation_grid, weight=5e2),
                                   RotationalTransform(eq=eq, bounds=(0.3, np.inf), grid=iota_grid, weight=5e3),
                                   BScaleLength(eq=eq, bounds=(0.2 + 0.02*(10-eq.NFP) - 0.004*aspect_ratio, np.inf), grid=Elongation_grid,\
                                                 normalize=False, normalize_target=False,  weight=scale_weight),
                                   QuasisymmetryTwoTerm(eq=eq, helicity=(1, 0), grid=QS_grid, weight=QS_weight)))
                                   #QuasisymmetryBoozer(eq=eq, helicity=(1, 0), grid=QS_grid, weight=QS_weight)))
    constraints = (
            ForceBalance(eq=eq),
            FixBoundaryR(eq=eq, modes=modes_R),
            FixBoundaryZ(eq=eq, modes=modes_Z),
            FixPressure(eq=eq),
            FixCurrent(eq=eq),
            FixPsi(eq=eq),
            )
    optimizer = Optimizer("proximal-lsq-exact")
    
    eq, _ = eq.optimize(
                objective,
                constraints,
                ftol=5e-4,
                xtol=1e-9,
                gtol=1e-9,
                maxiter=48,
                verbose=3
                )

    
    eq.save("eq_" + fname + f"_vacuum_{k:02d}_QS.h5")

    jax.clear_caches()



L0 = 10
rho = np.linspace(0, 1, L0+1)
grid = LinearGrid(L=L0, M = M_grid, N=N_grid, NFP=eq.NFP, sym=True)
data_keys = ["iota"]
iota = eq.compute(data_keys, grid=grid)["iota"]

eq.save("eq_" + fname + f"_vacuum_final2_QS.h5")

print("Final equilibrium nestedness = ", eq.is_nested())

fig, ax = plot_section(eq, "|F|", norm_F=True, log=True); 
plt.savefig(f"normF_vacuum_QS.png", dpi=400)  
plt.close()

eq0 = Equilibrium.load("eq_" + init_fname + "_vacuum_init_QS.h5")

fig, ax = plot_comparison(eqs=[eq0, eq]) 
plt.savefig(f"Xsection_vacuum_QS.png", dpi=400)  
plt.close()

fig, ax = plot_boozer_surface(eq, rho=1)
plt.savefig(f"Boozer_eq_vacuum_QS.png", dpi=400)  
plt.close()

```

**Comment by dpanici** [comment]
I will try running but do you have your own pics of the results? is it a stellarator QA or like does it become axisymmetric?

**Comment by dpanici** [comment]
<img width="291" height="275" alt="Image" src="https://github.com/user-attachments/assets/3a8f3067-07b1-4aa6-a321-2f22a87b2a2a" />

**Comment by dpanici** [comment]
f_C initial v final does decrease

The issue I think is that |B| decrease which decreases QS error. Try adding a aspect ratio of mean field strength objective

<img width="461" height="911" alt="Image" src="https://github.com/user-attachments/assets/5cf968e9-277b-4f48-b8f8-fd55af3a93af" />

**Comment by rahulgaur104** [comment]
I get the same plots. So the |B| strength goes down by a factor of 2 but the QS error that's printed goes down by 13 orders of magnitude without seeing any improvement in the quasisymmetry. Are the printed objective values completely nonsensical now? 

```
 19 Maximum absolute Quasi-symmetry (1,0) two-term error:        2.852e-02  -->   2.852e-02 (T^3)
 18 Minimum absolute Quasi-symmetry (1,0) two-term error:        2.218e-18  -->   2.218e-18 (T^3)
 17 Average absolute Quasi-symmetry (1,0) two-term error:        6.081e-03  -->   6.081e-03 (T^3)
 16 Maximum absolute Quasi-symmetry (1,0) two-term error:        1.674e-14  -->   1.674e-14 (normalized)
 15 Minimum absolute Quasi-symmetry (1,0) two-term error:        1.302e-30  -->   1.302e-30 (normalized)
 14 Average absolute Quasi-symmetry (1,0) two-term error:        3.570e-15  -->   3.569e-15 (normalized)
```

**Comment by dpanici** [comment]
Wait wdym 13 orders of magnitude?

**Comment by dpanici** [comment]
<img width="428" height="71" alt="Image" src="https://github.com/user-attachments/assets/04a74026-8cae-4059-a9c9-0e5d6b5c2eb4" />

**Comment by rahulgaur104** [comment]
Here's the full trace:

[slurm-43516160.txt](https://github.com/user-attachments/files/22684649/slurm-43516160.txt)

**Comment by dpanici** [comment]
<img width="595" height="673" alt="Image" src="https://github.com/user-attachments/assets/028ec8c2-0b5f-4302-a04c-7bc986fe61d7" />

**Comment by rahulgaur104** [comment]
Let me run `desc/examples/precise_QA.py`

**Comment by dpanici** [comment]
> > just that it was able to decrease QS error from the  initial point by making it more QA, even though you wanted QH. It did not really get "good" QH anyways.
> 
> I wanted QA. It looks nothing like a QA. If I wanted QH, I would've set the helicity=(1, NFP)

I mistyped in my initial post, I meant "wanted QA", i edited it now. the plot I shared was of f_C for QA, and it clearly shows it decreased QS error from the initial to the final.

> It's clearly a bug because the normalization is messed up and I end up with QH when I wanted QA. How sure are you that this normalization problem will not happen for any other equilibrium?

The normalizaiton is working as currently written, just the surface has gotten to some sort of regime where apparently the M=1 N=0 components are not contributing to the shaping meaningfully, which is kinda weird. That is a bug but is NOT affecting the first step of the optimization where it seems the majority of the optimzation steps were taking. So, the QA did in fact decrease (normalization issue notwithstanding) from the initial to the final (as you can see from the unnormalized f_C plots), so the optimizer worked as intended.

I agree yes bug in normalization but I argue there's no bug in the QS objective specifically.

> Also, here's the |F|_normalized. I don't know why you think it's decent.
> 
> <img alt="Image" width="2000" height="2000" src="https://private-user-images.githubusercontent.com/19224702/497238568-52c73080-d574-4d8a-8084-2c496ec90d83.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk1MTQwMTQsIm5iZiI6MTc1OTUxMzcxNCwicGF0aCI6Ii8xOTIyNDcwMi80OTcyMzg1NjgtNTJjNzMwODAtZDU3NC00ZDhhLTgwODQtMmM0OTZlYzkwZDgzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDAzVDE3NDgzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUwMTVhMzg2MjZlZDkyNjJiNjQxN2QyZjNhYzIyZTE5OGY5ZjcxMTBkOTgyNjNhY2M1ZmFkMzA3N2ViZGU5MWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Y-umw7zdmBIW5MYkKHDlEAkPvoKp9hNe8uz-o_K4VYA">

Ok yes it is definitely not good but what I meant was that is not the reason for the normalization going crazy or for the optimization to be completely off. Sorry for my wording there, you're right in that "decent" is not a good word for it, "not completely terrible" is maybe a better phrasing (like it is not unnested at some point)

Maybe also try solving to tighter than default solve tols to improve the force error?


So tl;dr the **QS objective isn't misbehaving**, it is just that given the Initial condition and the objective you constructed, it could in fact decrease QA error by making it appear more QH and by decreasing the magnitude of B, then somehow the normalization got messed up which made the last few steps not good and made f bal worse (f bal of the step-1 solution is good, like <1% everywhere, it is still slightly more "QH" looking despite having decreased f_C for QA during the optimization, but that is not a bug, at such high QS errors it is definitely possible that one can decrease QA error by making the contours visually more QH)

**Comment by dpanici** [comment]
for context, this is the eq surfaces, force balance, and boozer plot after step-one

So I think 
- the normalization getting messed up was what then caused most of the issues, as force balance is low here
- It was able to decrease QS error by some combination of straightening some contours (not to QA but still less closed contours than initially) and reducing the magnitude of |B| by lowering the aspect ratio (constant Psi, larger XS area = lower B), this is not a bug in the code just the optimizer doing its job to minimize the objective function given the constraints it had. 

Because our normalizations are not dynamically changing during optimization, this sort of thing can occur. We I think have discussed why we do this in the past, relating to not wanting to take a derivative through the normalization itself, so as long as we don't do dynamic normalization, we should be mindful to also constrain the problem to avoid "cheating" of the optimizer (in this case, set the aspect ratio target more tightly to the initial so it cannot vary the magnitude of the B field, and thus the other objectives, so much by just making the plasma XS larger)

<img width="906" height="778" alt="Image" src="https://github.com/user-attachments/assets/1818a4b0-42ec-40c0-bdda-2655f8bf47d6" />

<img width="488" height="461" alt="Image" src="https://github.com/user-attachments/assets/4ed34c87-631e-493d-ab22-48a26eb3c87f" />

<img width="1063" height="779" alt="Image" src="https://github.com/user-attachments/assets/46a05b6f-da15-4192-bb73-30ef307f08af" />
